xonsh/pyghooks.py

```diff
@@ -1373,54 +1375,79 @@ def color_file(file_path: str, mode: int) -> (Color, str):
-        color_key = "ln"
+    # if symlink, get info on (final) target
+    if stat.S_ISLNK(path_stat.st_mode):
         try:
-            os.stat(file_path)
-        except FileNotFoundError:
-            color_key = "or"
-    elif stat.S_ISREG(mode):
-        if stat.S_IMODE(mode) & (stat.S_IXUSR + stat.S_IXGRP + stat.S_IXOTH):
-            color_key = "ex"
-        elif (
-            mode & stat.S_ISUID
-        ):  # too many tests before we get to the common case -- restructure?
+            tar_path_stat = os.stat(file_path)  # and work with its properties
+            if lsc.is_target("ln"):  # if ln=target
+                path_stat = tar_path_stat
+        except FileNotFoundError:  # bug always color broken link 'or'
+            color_key = "or"  # early exit
+            ret_color_token = file_color_tokens.get(color_key, Text)
+            return ret_color_token, color_key
+
+    mode = path_stat.st_mode
+
+    if stat.S_ISREG(mode):
+        if mode & stat.S_ISUID:
             color_key = "su"
         elif mode & stat.S_ISGID:
             color_key = "sg"
         else:
-            match = color_file_extension_RE.match(file_path)
-            if match:
-                ext = "*" + match.group(1)  # look for *.<fileExtension> coloring
-                if ext in lsc:
-                    color_key = ext
-                else:
-                    color_key = "rs"
+            cap = os_listxattr(file_path, follow_symlinks=False)
+            if cap and "security.capability" in cap:  # protect None return on some OS?
+                color_key = "ca"
+            elif stat.S_IMODE(mode) & (stat.S_IXUSR + stat.S_IXGRP + stat.S_IXOTH):
+                color_key = "ex"
+            elif path_stat.st_nlink > 1:
+                color_key = "mh"
             else:
-                color_key = "rs"
-    elif stat.S_ISDIR(mode):  # ls -c doesn't colorize sticky or ow if not dirs...
-        color_key = ("di", "ow", "st", "tw")[
-            (mode & stat.S_ISVTX == stat.S_ISVTX) * 2
-            + (mode & stat.S_IWOTH == stat.S_IWOTH)
-        ]
-    elif stat.S_ISCHR(mode):
-        color_key = "cd"
-    elif stat.S_ISBLK(mode):
-        color_key = "bd"
+                color_key = "fi"
+    elif stat.S_ISDIR(mode):  # ls --color doesn't colorize sticky or ow if not dirs...
+        color_key = "di"
+        if (mode & stat.S_ISVTX) and (mode & stat.S_IWOTH):
+            color_key = "tw"
+        elif mode & stat.S_IWOTH:
+            color_key = "ow"
+        elif mode & stat.S_ISVTX:
+            color_key = "st"
+    elif stat.S_ISLNK(mode):
+        color_key = "ln"
     elif stat.S_ISFIFO(mode):
         color_key = "pi"
     elif stat.S_ISSOCK(mode):
         color_key = "so"
+    elif stat.S_ISBLK(mode):
+        color_key = "bd"
+    elif stat.S_ISCHR(mode):
+        color_key = "cd"
     elif stat.S_ISDOOR(mode):
-        color_key = "do"  # bug missing mapping for FMT based PORT and WHITEOUT ??
```

Comment:
Maybe this comment could be kept or replaced by another explaining that although S_ISPORT and S_ISWHT are defined in python's stat, our reference implementation (GNU's `ls`) does not use them (and as a consequence, no two-character type is defined for them).

Reply:
I will be happy to review your PR to add this support!

Reply:
I mean, there's not really anything *we* can do, because `ls` does not support ports/whiteouts, so there's no two-character type for them. I am not suggesting to add support, but simply to keep/replace the comment describing this fact.